syntax = "proto3";

package yandex.cloud.priv.iam.v1;

import "google/api/annotations.proto";
import "google/protobuf/timestamp.proto";
import "ydb/public/api/client/yc_private/iam/iam_token_service_subject.proto";
import "ydb/public/api/client/yc_private/iam/yandex_passport_cookie.proto";
import "ydb/public/api/client/yc_private/iam/oauth_request.proto";

service IamTokenService {

  // method doesn't require auth
  rpc Create (CreateIamTokenRequest) returns (CreateIamTokenResponse) {
    option (google.api.http) = { post: "/iam/v1/tokens" body: "*" };
  }

  // create iam token for service account
  rpc CreateForServiceAccount (CreateIamTokenForServiceAccountRequest) returns (CreateIamTokenResponse);

  // create iam token for service
  rpc CreateForService (CreateIamTokenForServiceRequest) returns (CreateIamTokenResponse);

  // create iam token for compute instance
  rpc CreateForComputeInstance (CreateIamTokenForComputeInstanceRequest) returns (CreateIamTokenResponse);

  // create iam token for oauth request
  rpc CreateForUserAccount (CreateIamTokenForSubjectRequest) returns (CreateIamTokenResponse) {
    option deprecated = true;
  }

  // create iam token for oauth request
  rpc CreateForSubject (CreateIamTokenForSubjectRequest) returns (CreateIamTokenResponse);
}

message CreateIamTokenRequest {
  oneof identity {
    string yandex_passport_oauth_token = 1;
    string jwt = 2;
    string iam_cookie = 3;
    YandexPassportCookies yandex_passport_cookies = 4;
  }
}

message CreateIamTokenForSubjectRequest {
  string subject_id = 1;
  oauth.v1.OAuthRequest oauth_request = 2;
  // additional parameters for cross-domain authentication
  string session_id = 3;
}

message CreateIamTokenForServiceAccountRequest {
  string service_account_id = 1;
}

message CreateIamTokenForServiceRequest {
  string service_id = 1;
  string microservice_id = 2;
  string resource_id = 3;
  string resource_type = 4;
  string target_service_account_id = 5;
}

message CreateIamTokenForComputeInstanceRequest {
  string service_account_id = 1;
  string instance_id = 2;
}

message CreateIamTokenResponse {
  string iam_token = 1;
  google.protobuf.Timestamp issued_at = 4;
  google.protobuf.Timestamp expires_at = 2;
  ts.Subject subject = 3;
}
